### PR TITLE
#3499 Ledger back end query

### LINKED ISSUE #3499: Back end history

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #3499

# 👩🏻‍💻 What does this PR do? 
 

Just putting this here in Draft initially so you can see if I've missed anything obvious.

Filtering by StockLineId, sorting by most fields, no pagination for now.

Note: this is implemented in #3588 now, so you can use that branch to test if you want.

# 🧪 How has/should this change been tested? 

Queried "Ledger" endpoint using GraphQL browser, filter by StocklineId, can sort by most fields.

## 💌 Any notes for the reviewer?

What additional tests do you think we need to add?

### CODE REVIEW

**Review by lache-melvin — COMMENTED** [review]
Looking good! Curious what your plan is with the pagination? I think if not implementing, the pattern we have elsewhere is to pass `Pagination::one()` to pass to the database query, rather than not include at all (just don't expose the input from graphql)

File: server/graphql/general/src/lib.rs [hunk 1]
```diff
@@ -143,6 +143,18 @@ impl GeneralQueries {
...
+    pub async fn ledger(
+        &self,
+        ctx: &Context<'_>,
+        store_id: String,
+        // #[graphql(desc = "Pagination option (first and offset)")] page: Option<PaginationInput>,
+        #[graphql(desc = "Filter option")] filter: Option<LedgerFilterInput>,
+        #[graphql(desc = "Sort options (only first sort input is evaluated for this endpoint)")]
+        sort: Option<Vec<LedgerSortInput>>,
+    ) -> Result<LedgerResponse> {
+        ledger(ctx, store_id, /*page,  */ filter, sort)
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/graphql/general/src/lib.rs:155]: Not planning to paginate at this point. Will do the `Pagination::one()` as suggested :)

File: server/graphql/general/src/queries/ledger.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+use service::{
+    auth::{Resource, ResourceAccessRequest},
+    ledger::get_ledger,
+    ListResult,
+};
+
+#[derive(Enum, Copy, Clone, PartialEq, Eq)]
+#[graphql(rename_items = "camelCase")]
+pub enum LedgerSortFieldInput {
+    Id,
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/graphql/general/src/queries/ledger.rs]: Sure, have removed. fc41e892e

File: server/graphql/general/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/graphql/general/src/lib.rs]: fc41e892e

**Comment by lache-melvin** [comment]
When I run the migrations with Postgres, I get this error: `"UNKNOWN" ("\"\\\"column invoice_line_stock_movement.return_reason_id does not exist\\\"\"")`

I assume this is because the `invoice_line_stock_movement` was most recently recreated in the 1.7 migrations, before the return_reason_id column existed - so probably need to recreate this view too?

File: server/service/src/ledger.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+pub fn get_ledger(
+    connection_manager: &StorageConnectionManager,
+    filter: Option<LedgerFilter>,
+    sort: Option<LedgerSort>,
+) -> Result<ListResult<LedgerRow>, ListError> {
+    // let pagination = get_default_pagination(pagination, MAX_LIMIT, MIN_LIMIT)?;
+    let connection = connection_manager.connection()?;
+    let repository = LedgerRepository::new(&connection);
+
+    let rows = repository.query(Pagination::one(), filter.clone(), sort)?;
```
Comment (lache-melvin) [inline server/service/src/ledger.rs:22]: Oh hahaha I meant `Pagination::all()`, not `one()`, sorry! Now this only getting one line 😅

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Looks like its working great with SQLite Carl, just need to sort for PG too 🙏 

Also sorry about the wrong point on the Pagination::all() lol

File: server/service/src/ledger.rs (hunk 3 above)
Comment (CarlosNZ) [reply server/service/src/ledger.rs]: Ha, oops, okay. 6f08d5328

**Comment by CarlosNZ** [comment]
> Looks like its working great with SQLite Carl, just need to sort for PG too 🙏

Ah yes, looked at this with @andreievg, came to the conclusion that using `invoice.*` in the existing "invoice_line_stock_movement" was causing problems. We figured the cleanest thing to do was to drop and re-create all the related views (plus that gives us a chance to get rid of three of them that are no longer required).

Have added migrations, plus explanatory comment: 8b8b27e1f

**Review by lache-melvin — APPROVED** [review]
Looks great! Thanks for changes 🙏

**Review by lache-melvin — CHANGES_REQUESTED** [review]
Wait kidding the tests are failing to compile... 😅

**Comment by lache-melvin** [comment]
Oh if I get the tests compiling, this one fails... possibly something missing in the new view?

Screenshot 2024-04-23 at 9 04 24 AM

**Comment by CarlosNZ** [comment]
> Oh if I get the tests compiling, this one fails... possibly something missing in the new view?

I made the fix to test in db_diesel and that one's working for me now. Did you have another particular test that is still failing?

**Comment by lache-melvin** [comment]
> > Oh if I get the tests compiling, this one fails... possibly something missing in the new view?
> 
> I made the fix to test in db_diesel and that one's working for me now. Did you have another particular test that is still failing?

Oh this conversation moved to Telegram but for visibility - failing test is `test_graphql_item_stats_loader` - looks like something has changed with the consumption view 👀

**Comment by CarlosNZ** [comment]
@lache-melvin Tests all working now. 

The file `[server/repository/src/migrations/v2_01_00/ledger.rs](https://github.com/msupply-foundation/open-msupply/pull/3639/files#diff-4d94cc4b6d281008104c73d259668fa7e10b0b4ab78821658f5a55ef624797a3)` contains the changes I did to get it working:

- Remove the "stockline_id IS NOT NULL" condition for "stock_movement" View (to make it equivalent to previous version)
- Update diesel table definition for above to reflect "stockline_id" now being Nullable, and related types

**Review by lache-melvin — APPROVED** [review]
Looks great! Thanks for resolving 🙏

**Pull request merged**